File: packages/sdk-flutter/lib/src/analytics/event_queue.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/foundation.dart';
import 'package:hive/hive.dart';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';

import '../models/event.dart';

/// Manages event queuing with offline support.
///
/// Features:
/// - Stores events locally when offline
/// - Automatically sends events when network is restored
/// - Batches events for efficient transmission
/// - Implements exponential backoff on failure
/// - Persists events across app restarts
class EventQueue {
  EventQueue._();

  static EventQueue? _instance;

  /// Singleton instance
  static EventQueue get instance {
    _instance ??= EventQueue._();
    return _instance!;
  }

  // Configuration
  String? _apiEndpoint;
  String? _apiKey;
  bool _debug = false;
  http.Client? _httpClient;

  // Hive storage
  Box<String>? _eventBox;
  static const String _boxName = 'eodin_events';

  // Queue settings
  static const int maxQueueSize = 1000;
  static const int maxBatchSize = 50;
  static const int flushThreshold = 20;
  static const Duration flushInterval = Duration(seconds: 30);
  static const Duration maxEventAge = Duration(days: 7);

  // State
  final List<AnalyticsEvent> _memoryQueue = [];
  bool _isInitialized = false;
  bool _isFlushing = false;
  Timer? _flushTimer;
  int _retryCount = 0;
  static const int maxRetries = 5;

  // Network monitoring
  StreamSubscription<List<ConnectivityResult>>? _connectivitySubscription;
  bool _isOnline = true;

  /// Initialize the event queue
  Future<void> initialize({
    required String apiEndpoint,
    required String apiKey,
    bool debug = false,
    http.Client? httpClient,
  }) async {
    if (_isInitialized) return;

    _apiEndpoint = apiEndpoint;
    _apiKey = apiKey;
    _debug = debug;
    _httpClient = httpClient;

    // Initialize Hive
    try {
      final appDocDir = await getApplicationDocumentsDirectory();
      Hive.init('${appDocDir.path}/eodin_analytics');
      _eventBox = await Hive.openBox<String>(_boxName);
      _log('Hive initialized with ${_eventBox!.length} stored events');
    } catch (e) {
      _log('Failed to initialize Hive: $e', isError: true);
    }

    // Load persisted events
    await _loadPersistedEvents();

    // Start network monitoring
    _startNetworkMonitoring();

    // Start flush timer
    _startFlushTimer();

    _isInitialized = true;
    _log('EventQueue initialized');
  }

  /// Add an event to the queue
  Future<void> enqueue(AnalyticsEvent event) async {
    if (!_isInitialized) {
      _log('EventQueue not initialized', isError: true);
      return;
    }

    // Check queue size limit
    if (_memoryQueue.length >= maxQueueSize) {
      // Remove oldest events
      final removeCount = _memoryQueue.length - maxQueueSize + 1;
      _memoryQueue.removeRange(0, removeCount);
      _log('Queue overflow, removed $removeCount old events');
    }

    _memoryQueue.add(event);
    _log('Enqueued event: ${event.eventName} (queue size: ${_memoryQueue.length})');

    // Persist to Hive
    await _persistEvent(event);

    // Check if we should flush
    if (_isOnline && _memoryQueue.length >= flushThreshold) {
      await flush();
    }
  }

  /// Flush pending events to the server
  Future<void> flush() async {
    if (!_isInitialized || _isFlushing || _memoryQueue.isEmpty) {
      return;
    }

    if (!_isOnline) {
      _log('Offline, skipping flush');
      return;
    }

    _isFlushing = true;

    try {
      // Take up to maxBatchSize events
      final eventsToSend = _memoryQueue.take(maxBatchSize).toList();

      _log('Flushing ${eventsToSend.length} events');

      final success = await _sendEvents(eventsToSend);

      if (success) {
        // Remove sent events from memory queue
        _memoryQueue.removeRange(0, eventsToSend.length);

        // Remove from Hive storage
        for (final event in eventsToSend) {
          await _removePersistedEvent(event.eventId);
        }

        _retryCount = 0;
        _log('Flush successful');

        // Continue flushing if more events
        if (_memoryQueue.isNotEmpty) {
          await flush();
        }
      } else {
        _retryCount++;
        _log('Flush failed, retry count: $_retryCount', isError: true);

        // Schedule retry with exponential backoff
        if (_retryCount <= maxRetries) {
          final delay = _calculateBackoff(_retryCount);
          _log('Scheduling retry in ${delay.inSeconds}s');
          Timer(delay, () => flush());
        }
      }
    } finally {
      _isFlushing = false;
    }
  }

  /// Get current queue size
  int get queueSize => _memoryQueue.length;

  /// Get persisted event count
  int get persistedEventCount => _eventBox?.length ?? 0;

  /// Whether currently online
  bool get isOnline => _isOnline;

  /// Dispose resources
  Future<void> dispose() async {
    _flushTimer?.cancel();
    _connectivitySubscription?.cancel();
    await _eventBox?.close();
    _isInitialized = false;
    _log('EventQueue disposed');
  }

  // Private methods

  Future<bool> _sendEvents(List<AnalyticsEvent> events) async {
    if (_apiEndpoint == null || _apiKey == null) {
      return false;
    }

    final client = _httpClient ?? http.Client();

    try {
      final response = await client.post(
        Uri.parse('$_apiEndpoint/events/collect'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          'X-API-Key': _apiKey!,
        },
        body: jsonEncode({
          'events': events.map((e) => e.toJson()).toList(),
        }),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final json = jsonDecode(response.body);
        _log('Send successful: received=${json['received']}, duplicates=${json['duplicates']}');
        return true;
      } else {
        _log('Send failed: ${response.statusCode} - ${response.body}', isError: true);
        return false;
      }
    } on SocketException catch (e) {
      _log('Network error: $e', isError: true);
      _isOnline = false;
      return false;
    } on TimeoutException catch (e) {
      _log('Timeout error: $e', isError: true);
      return false;
    } catch (e) {
      _log('Send error: $e', isError: true);
      return false;
    } finally {
      if (_httpClient == null) {
        client.close();
      }
    }
  }

  Future<void> _persistEvent(AnalyticsEvent event) async {
    if (_eventBox == null) return;

    try {
      final jsonStr = jsonEncode(event.toJson());
      await _eventBox!.put(event.eventId, jsonStr);
    } catch (e) {
      _log('Failed to persist event: $e', isError: true);
    }
  }

  Future<void> _removePersistedEvent(String eventId) async {
    if (_eventBox == null) return;

    try {
      await _eventBox!.delete(eventId);
    } catch (e) {
      _log('Failed to remove persisted event: $e', isError: true);
    }
  }

  Future<void> _loadPersistedEvents() async {
    if (_eventBox == null) return;

    final now = DateTime.now();
    final expiredKeys = <String>[];

    for (final key in _eventBox!.keys) {
      try {
        final jsonStr = _eventBox!.get(key);
        if (jsonStr == null) continue;

        final json = jsonDecode(jsonStr) as Map<String, dynamic>;
        final event = _eventFromJson(json);

        // Check if event is too old
        if (now.difference(event.timestamp) > maxEventAge) {
          expiredKeys.add(key as String);
          continue;
        }

        // Check if already in memory queue
        if (!_memoryQueue.any((e) => e.eventId == event.eventId)) {
          _memoryQueue.add(event);
        }
      } catch (e) {
        _log('Failed to load persisted event: $e', isError: true);
        expiredKeys.add(key as String);
      }
    }

    // Remove expired events
    for (final key in expiredKeys) {
      await _eventBox!.delete(key);
    }

    _log('Loaded ${_memoryQueue.length} events from storage, removed ${expiredKeys.length} expired');
  }

  AnalyticsEvent _eventFromJson(Map<String, dynamic> json) {
    return AnalyticsEvent(
      eventId: json['event_id'] as String,
      eventName: json['event_name'] as String,
      appId: json['app_id'] as String,
      deviceId: json['device_id'] as String,
      userId: json['user_id'] as String?,
      sessionId: json['session_id'] as String?,
      timestamp: DateTime.parse(json['timestamp'] as String),
      attribution: json['attribution'] != null
          ? Attribution.fromJson(json['attribution'] as Map<String, dynamic>)
          : null,
      device: json['device'] != null
          ? _deviceInfoFromJson(json['device'] as Map<String, dynamic>)
          : null,
      properties: json['properties'] as Map<String, dynamic>?,
    );
  }

  DeviceInfo _deviceInfoFromJson(Map<String, dynamic> json) {
    return DeviceInfo(
      os: json['os'] as String?,
      osVersion: json['os_version'] as String?,
      model: json['model'] as String?,
      locale: json['locale'] as String?,
      attStatus: json['att_status'] as String?,
      idfa: json['idfa'] as String?,
    );
  }

  void _startNetworkMonitoring() {
    _connectivitySubscription = Connectivity()
        .onConnectivityChanged
        .listen((List<ConnectivityResult> results) {
      final wasOffline = !_isOnline;
      _isOnline = results.isNotEmpty &&
          !results.contains(ConnectivityResult.none);

      _log('Connectivity changed: $_isOnline');

      // If we just came online, flush events
      if (wasOffline && _isOnline && _memoryQueue.isNotEmpty) {
        _log('Network restored, flushing events');
        flush();
      }
    });

    // Check initial connectivity
    Connectivity().checkConnectivity().then((results) {
      _isOnline = results.isNotEmpty &&
          !results.contains(ConnectivityResult.none);
      _log('Initial connectivity: $_isOnline');
    });
  }

  void _startFlushTimer() {
    _flushTimer?.cancel();
    _flushTimer = Timer.periodic(flushInterval, (_) {
      if (_isOnline && _memoryQueue.isNotEmpty) {
        flush();
      }
    });
  }

  Duration _calculateBackoff(int retryCount) {
    // Exponential backoff: 1s, 2s, 4s, 8s, 16s (capped at 60s)
    final seconds = min(pow(2, retryCount).toInt(), 60);
    return Duration(seconds: seconds);
  }

  void _log(String message, {bool isError = false}) {
    if (_debug || kDebugMode) {
      if (isError) {
        debugPrint('[EodinEventQueue ERROR] $message');
      } else {
        debugPrint('[EodinEventQueue] $message');
      }
    }
  }

  /// Reset for testing
  @visibleForTesting
  Future<void> reset() async {
    _flushTimer?.cancel();
    _connectivitySubscription?.cancel();
    _memoryQueue.clear();
    _isInitialized = false;
    _isFlushing = false;
    _retryCount = 0;

    if (_eventBox != null) {
      await _eventBox!.clear();
      await _eventBox!.close();
      _eventBox = null;
    }
  }
}

File: packages/capacitor/ios/Sources/EodinCapacitorPlugin/EventQueue.swift
import Foundation
#if canImport(Network)
import Network
#endif

/// Manages event queuing with offline support.
///
/// Features:
/// - Stores events locally when offline
/// - Automatically sends events when network is restored
/// - Batches events for efficient transmission
/// - Implements exponential backoff on failure
/// - Persists events across app restarts
public final class EventQueue {

    // MARK: - Singleton

    /// Shared instance
    public static let shared = EventQueue()

    private init() {}

    // MARK: - Configuration

    private var apiEndpoint: String?
    private var apiKey: String?
    private var isDebug = false

    // MARK: - Constants

    private let maxQueueSize = 1000
    private let maxBatchSize = 50
    private let flushThreshold = 20
    private let flushInterval: TimeInterval = 30
    private let maxEventAge: TimeInterval = 7 * 24 * 60 * 60 // 7 days
    private let maxRetries = 5

    // MARK: - State

    private var memoryQueue: [AnalyticsEvent] = []
    private var isInitialized = false
    private var isFlushing = false
    private var retryCount = 0
    private var flushTimer: Timer?

    // MARK: - Storage

    private let storageKey = "eodin_event_queue"
    private let queue = DispatchQueue(label: "app.eodin.eventqueue", qos: .utility)

    // MARK: - Network Monitoring

    #if canImport(Network)
    private var pathMonitor: NWPathMonitor?
    #endif
    private(set) var isOnline = true

    // MARK: - Public Methods

    /// Initialize the event queue
    public func initialize(
        apiEndpoint: String,
        apiKey: String,
        debug: Bool = false
    ) {
        guard !isInitialized else { return }

        self.apiEndpoint = apiEndpoint
        self.apiKey = apiKey
        self.isDebug = debug

        // Load persisted events
        loadPersistedEvents()

        // Start network monitoring
        startNetworkMonitoring()

        // Start flush timer
        startFlushTimer()

        isInitialized = true
        log("EventQueue initialized with \(memoryQueue.count) persisted events")
    }

    /// Add an event to the queue
    public func enqueue(_ event: AnalyticsEvent) {
        queue.async { [weak self] in
            guard let self = self, self.isInitialized else { return }

            // Check queue size limit
            if self.memoryQueue.count >= self.maxQueueSize {
                let removeCount = self.memoryQueue.count - self.maxQueueSize + 1
                self.memoryQueue.removeFirst(removeCount)
                self.log("Queue overflow, removed \(removeCount) old events")
            }

            self.memoryQueue.append(event)
            self.log("Enqueued event: \(event.eventName) (queue size: \(self.memoryQueue.count))")

            // Persist to storage
            self.persistEvents()

            // Check if we should flush
            if self.isOnline && self.memoryQueue.count >= self.flushThreshold {
                self.performFlush()
            }
        }
    }

    /// Flush pending events to the server
    public func flush() {
        queue.async { [weak self] in
            self?.performFlush()
        }
    }

    /// Current queue size
    public var queueSize: Int {
        return memoryQueue.count
    }

    /// Dispose resources
    public func dispose() {
        flushTimer?.invalidate()
        flushTimer = nil

        #if canImport(Network)
        pathMonitor?.cancel()
        pathMonitor = nil
        #endif

        isInitialized = false
        log("EventQueue disposed")
    }

    // MARK: - Private Methods

    private func performFlush() {
        guard isInitialized, !isFlushing, !memoryQueue.isEmpty else { return }
        guard isOnline else {
            log("Offline, skipping flush")
            return
        }

        isFlushing = true

        // Take up to maxBatchSize events
        let eventsToSend = Array(memoryQueue.prefix(maxBatchSize))

        log("Flushing \(eventsToSend.count) events")

        sendEvents(eventsToSend) { [weak self] success in
            guard let self = self else { return }

            self.queue.async {
                if success {
                    // Remove sent events
                    self.memoryQueue.removeFirst(min(eventsToSend.count, self.memoryQueue.count))
                    self.persistEvents()
                    self.retryCount = 0
                    self.log("Flush successful")

                    // Continue flushing if more events
                    self.isFlushing = false
                    if !self.memoryQueue.isEmpty {
                        self.performFlush()
                    }
                } else {
                    self.retryCount += 1
                    self.log("Flush failed, retry count: \(self.retryCount)")

                    self.isFlushing = false

                    // Schedule retry with exponential backoff
                    if self.retryCount <= self.maxRetries {
                        let delay = self.calculateBackoff(self.retryCount)
                        self.log("Scheduling retry in \(Int(delay))s")

                        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
                            self?.flush()
                        }
                    }
                }
            }
        }
    }

    private func sendEvents(_ events: [AnalyticsEvent], completion: @escaping (Bool) -> Void) {
        guard let endpoint = apiEndpoint, let apiKey = apiKey else {
            completion(false)
            return
        }

        guard let url = URL(string: "\(endpoint)/events/collect") else {
            completion(false)
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.timeoutInterval = 30

        do {
            let body = ["events": events]
            request.httpBody = try JSONEncoder().encode(body)
        } catch {
            log("Failed to encode events: \(error)")
            completion(false)
            return
        }

        let task = URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                self?.log("Send error: \(error)")
                completion(false)
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                completion(false)
                return
            }

            if httpResponse.statusCode == 200 {
                if let data = data,
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    self?.log("Send successful: received=\(json["received"] ?? "?"), duplicates=\(json["duplicates"] ?? "?")")
                }
                completion(true)
            } else {
                self?.log("Send failed: \(httpResponse.statusCode)")
                completion(false)
            }
        }

        task.resume()
    }

    private func persistEvents() {
        do {
            let eventsData = try JSONEncoder().encode(memoryQueue)
            UserDefaults.standard.set(eventsData, forKey: storageKey)
        } catch {
            log("Failed to persist events: \(error)")
        }
    }

    private func loadPersistedEvents() {
        guard let data = UserDefaults.standard.data(forKey: storageKey) else { return }

        do {
            var events = try JSONDecoder().decode([AnalyticsEvent].self, from: data)

            // Remove expired events
            let now = Date()
            events = events.filter { now.timeIntervalSince($0.timestamp) < maxEventAge }

            memoryQueue = events
            log("Loaded \(events.count) persisted events")
        } catch {
            log("Failed to load persisted events: \(error)")
            UserDefaults.standard.removeObject(forKey: storageKey)
        }
    }

    private func startNetworkMonitoring() {
        #if canImport(Network)
        if #available(iOS 12.0, *) {
            pathMonitor = NWPathMonitor()
            pathMonitor?.pathUpdateHandler = { [weak self] path in
                let wasOffline = !(self?.isOnline ?? true)
                self?.isOnline = path.status == .satisfied

                self?.log("Network status changed: \(self?.isOnline == true ? "online" : "offline")")

                // If we just came online, flush events
                if wasOffline && self?.isOnline == true && !(self?.memoryQueue.isEmpty ?? true) {
                    self?.log("Network restored, flushing events")
                    self?.flush()
                }
            }
            pathMonitor?.start(queue: queue)
        }
        #endif
    }

    private func startFlushTimer() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            self.flushTimer?.invalidate()
            self.flushTimer = Timer.scheduledTimer(withTimeInterval: self.flushInterval, repeats: true) { [weak self] _ in
                guard let self = self else { return }
                if self.isOnline && !self.memoryQueue.isEmpty {
                    self.flush()
                }
            }
        }
    }

    private func calculateBackoff(_ retryCount: Int) -> TimeInterval {
        // Exponential backoff: 1s, 2s, 4s, 8s, 16s (capped at 60s)
        return min(pow(2.0, Double(retryCount)), 60.0)
    }

    private func log(_ message: String) {
        #if DEBUG
        if isDebug {
            print("[EodinEventQueue] \(message)")
        }
        #endif
    }

    // MARK: - Testing Support

    @available(*, deprecated, message: "For testing only")
    public func reset() {
        flushTimer?.invalidate()
        flushTimer = nil

        #if canImport(Network)
        pathMonitor?.cancel()
        pathMonitor = nil
        #endif

        memoryQueue.removeAll()
        UserDefaults.standard.removeObject(forKey: storageKey)
        isInitialized = false
        isFlushing = false
        retryCount = 0
    }
}
